### PR TITLE
Add a simple oak-functions loader

### LINKED ISSUE #1923: Simple "Oak Functions" loader and wasm engine
Related to #1919:
Implement a simple HTTP server that spawns a new wasm engine for every request and runs a very simple input wasm module.

### PR DESCRIPTION
**Pull request by rbehjati**
Fixes #1923.

# Checklist

- [ ] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [ ] I have written tests that cover the code changes.
  - [ ] I have checked that these tests are run by
        [Cloudbuild](/cloudbuild.yaml)
  - [ ] I have updated [documentation](/docs/) accordingly.
  - [ ] I have raised an [issue](https://github.com/project-oak/oak/issues) to
        cover any TODOs and/or unfinished work.
- [ ] Pull request includes prototype/experimental work that is under
      construction.

### CODE REVIEW

File: oak_functions/loader/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+[features]
+# Feature that allows logging. Should be use only for debugging purposes.
+oak-unsafe = []
+default = []
+
+[dependencies]
+anyhow = "*"
+env_logger = "*"
+http = "*"
+hyper = { version = "*", features = ["full"] }
```
Comment (tiziano88) [inline oak_functions/loader/Cargo.toml:19]: Is there any chance we can have less than `full` here?

File: oak_functions/loader/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+#[cfg(test)]
+mod tests;
+
+/// Command line options for the Oak loader.
+#[derive(StructOpt, Clone, Debug)]
+#[structopt(about = "Oak Functions Loader")]
+pub struct Opt {
+    #[structopt(long, default_value = "8080", help = "Server Port number.")]
+    server_port: u16,
```
Comment (tiziano88) [inline oak_functions/loader/src/main.rs:31]: Should this be `http_listen_address` or similar? I thought we use something like that elsewhere, also it seems more explicit.

File: oak_functions/loader/src/server.rs [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+#[derive(Clone)]
+struct WasmEngine {
+    // wasm module to be served on each invocation.
+    module: Arc<wasmi::Module>,
+}
+
+impl WasmEngine {
+    async fn handle_request(&self, req: Request<Body>) -> anyhow::Result<Response<Body>> {
```
Comment (tiziano88) [inline oak_functions/loader/src/server.rs:42]: Could you add a simple benchmark for this? I would be curious to see how long it actually takes to start the new wasm engine each time. Can be a separate PR.

File: oak_functions/loader/src/server.rs [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+use log::info;
+use std::{fs, net::SocketAddr, sync::Arc};
+
+const MAIN_FUNCTION_NAME: &str = "main";
+
+/// An HTTP server that serves the `main` function of a wasm module on each invocation.
+pub struct WasmServer {
+    /// Server address to listen for client requests on.
+    address: SocketAddr,
+    /// The wasm engine that runs the given wasm module per incoming request.
```
Comment (tiziano88) [inline oak_functions/loader/src/server.rs:31]:
```suggestion
    /// The Wasm engine that runs the given wasm module per incoming request.
```

and elsewhere (see https://webassembly.org/)

File: oak_functions/loader/src/server.rs [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        // TODO(#1919): Get the actual response from the wasm module via ABI functions.
+        http::response::Builder::new()
+            .status(StatusCode::BAD_REQUEST)
+            .body(Body::from("Not implemented yet.\n"))
+            .context("Couldn't create response")
+    }
+}
+
+impl WasmServer {
+    /// Create a WasmServer to listen on the given port, serving the given wasm module.
```
Comment (tiziano88) [inline oak_functions/loader/src/server.rs:67]:
```suggestion
    /// Creates a [`WasmServer`] instance listening on the given port, serving the given wasm module.
```

File: oak_functions/loader/src/server.rs [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+/// An HTTP server that serves the `main` function of a wasm module on each invocation.
+pub struct WasmServer {
+    /// Server address to listen for client requests on.
+    address: SocketAddr,
+    /// The wasm engine that runs the given wasm module per incoming request.
+    wasm_engine: WasmEngine,
+}
+
+#[derive(Clone)]
+struct WasmEngine {
```
Comment (tiziano88) [inline oak_functions/loader/src/server.rs:36]: Should we call this `WasmHandler` or something like that? I think that makes it clearer that it is ephemeral and corresponds to a request handler.

File: oak_functions/loader/src/server.rs [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+            .body(Body::from("Not implemented yet.\n"))
+            .context("Couldn't create response")
+    }
+}
+
+impl WasmServer {
+    /// Create a WasmServer to listen on the given port, serving the given wasm module.
+    pub fn create(address: &str, wasm_path: &str) -> anyhow::Result<Self> {
+        let wasm_module_bytes =
+            fs::read(&wasm_path).with_context(|| format!("Couldn't read file {}", wasm_path))?;
```
Comment (tiziano88) [inline oak_functions/loader/src/server.rs:70]: I think it would be more decoupled if the caller (in our case main) did the file reading, and this fn directly got the bytes of the wasm module to instantiate. It will also make testing easier later on.

File: oak_functions/loader/src/tests.rs [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        WasmServer::create(&address, TEST_WASM_MODULE_PATH).expect("Couldn't create the server");
+
+    let server_fut = server.start(notify_receiver);
+    let client_fut = create_client(port, notify_sender);
+
+    let (res, _) = tokio::join!(server_fut, client_fut);
+    assert!(res.is_ok());
+}
+
+async fn create_client(port: u16, notify_sender: tokio::sync::oneshot::Sender<()>) {
```
Comment (tiziano88) [inline oak_functions/loader/src/tests.rs:38]: This is not just creating the client, it is also making a request, right?

**Review by tiziano88 — COMMENTED** [review]
Looks great already! Thanks!

**Review by ipetr0v — APPROVED** [review]

File: oak_functions/loader/src/main.rs [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    wasm_path: String,
+}
+
+/// Main execution point for the Oak Functions Loader.
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+    if cfg!(feature = "oak-unsafe") {
+        env_logger::init();
+    } else {
+        eprintln!("No debugging output configured at build time");
```
[the remark this answers is not in the document]
Comment (tiziano88) [reply oak_functions/loader/src/main.rs:45]: there is no logger installed in this branch :)

File: oak_functions/loader/src/server.rs (hunk 3 above)
Comment (rbehjati) [reply oak_functions/loader/src/server.rs]: I have added a benchmark test, using Rust `Bencher`. This is currently not being run by the runner. I can fix that in a separate PR. 

Here is the output: 
```
test tests::bench_wasm_handler ... bench:      25,608 ns/iter (+/- 976)
```
Is this the type of test you were thinking of?

**Review by rbehjati — COMMENTED** [review]
Thanks for the reviews.

File: oak_functions/loader/Cargo.toml (hunk 1 above)
Comment (tiziano88) [reply oak_functions/loader/Cargo.toml]: AFAICT client is only needed for tests, so maybe it should be in dev-deps?

File: oak_functions/loader/src/server.rs (hunk 3 above)
Comment (tiziano88) [reply oak_functions/loader/src/server.rs]: Nice! That time seems low compared to what I expected, so I guess that's good news. (I expected ~100ms roughly). Adding it to runner later on sounds good, although we should think about how to make sure we don't regress (just running the bench does not do much, it can easily climb up to 10s per request, and nothing will fail, unless we add additional logic to check that the time stays below a given threshold, e.g. 50ms for this function. Maybe Bencher already does this?

Comment (tiziano88) [reply oak_functions/loader/src/server.rs]: BTW this crate seems to do something similar to what I'm thinking: https://bheisler.github.io/criterion.rs/book/user_guide/command_line_options.html#baselines

File: oak_functions/loader/src/server.rs [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+use log::info;
+use std::{net::SocketAddr, sync::Arc};
+
+const MAIN_FUNCTION_NAME: &str = "main";
+
+// An ephemeral request handler with a Wasm module for handling the requests.
+#[derive(Clone)]
+pub(crate) struct WasmHandler {
+    // Wasm module to be served on each invocation.
+    module: Arc<wasmi::Module>,
```
Comment (tiziano88) [inline oak_functions/loader/src/server.rs:31]: Not sure I see why this needs to be an `Arc` instead of just `wasmi::Module`. Is it because of the lifetimes involved in `wasmi::ModuleInstance::new`? If so, could you add a comment?

File: oak_functions/loader/src/server.rs [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+/// An HTTP server that serves the `main` function of a Wasm module on each invocation.
+pub struct WasmServer {
+    /// Server address to listen for client requests on.
+    address: SocketAddr,
+    /// The Wasm handler that runs the given Wasm module per incoming request.
+    wasm_handler: WasmHandler,
+}
+
+impl WasmServer {
+    /// Creates a [`WasmServer`] instance listening on the given port, serving the given Wasm
```
Comment (tiziano88) [inline oak_functions/loader/src/server.rs:74]: I don't think this actually starts listening on the given port / address, though maybe `create` and `start` could be combined in one (up to you); or is there anything useful that a non-started `WasmServer` can be used for?

File: oak_functions/loader/src/tests.rs [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+use hyper::client::Client;
+
+use std::fs;
+
+const TEST_WASM_MODULE_PATH: &str = "testdata/non-oak-minimal.wasm";
+
+#[tokio::test]
+async fn test_server() {
+    let _ = env_logger::builder().is_test(true).try_init();
+    let port = 9999;
```
Comment (tiziano88) [inline oak_functions/loader/src/tests.rs:26]: I guess at some point we will want to randomize this.

File: oak_functions/loader/src/tests.rs [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+
+    // TODO(#1919): Check the response.
+
+    notify_sender
+        .send(())
+        .expect("Couldn't send completion signal.");
+}
+
+extern crate test;
+use test::Bencher;
```
Comment (tiziano88) [inline oak_functions/loader/src/tests.rs:62]: Group all the `use` together?

**Review by tiziano88 — APPROVED** [review]

File: oak_functions/loader/src/main.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (tiziano88) [reply oak_functions/loader/src/main.rs:54]: `eprintln` is correct, we want to print to stderr, not stdout, right?

**Review by rbehjati — COMMENTED** [review]
Thanks!

File: oak_functions/loader/src/server.rs (hunk 3 above)
Comment (rbehjati) [reply oak_functions/loader/src/server.rs]: That crate looks interesting. But I think for that to work on the CI builds we have to tweak some GitHub action scripts. For now I am calling `bench` directly on the `bencher` to be able to get the results and have assertions. I have also created #1933 to track benchmarking.

File: oak_functions/loader/src/main.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (tiziano88) [reply oak_functions/loader/src/main.rs:54]: Well, this line is not the "output" of the program, it is just diagnostic, the same as a `log::info` basically, right? I think in general the kind of programs we usually write don't have any output per se, so I don't expect them to ever print anything to stdout.

File: oak_functions/loader/src/tests.rs [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+#[bench]
+fn bench_wasm_handler(bencher: &mut Bencher) {
+    let summary = bencher.bench(|bencher| {
+        let wasm_module_bytes =
+            fs::read(TEST_WASM_MODULE_PATH).expect("Couldn't read test Wasm module");
+        let wasm_handler =
+            WasmHandler::create(&wasm_module_bytes).expect("Couldn't create the server");
+        let request = hyper::Request::builder()
+            .method(http::Method::GET)
+            .uri("http://localhost:8080")
```
Comment (tiziano88) [inline oak_functions/loader/src/tests.rs:73]: I was a bit confused seeing this line, I thought it was doing an actual request to this endpoint, while actually I guess it is just to make the request build, right? If so, maybe change it to example.com or something more obviously unrelated to the rest of the code? up to you

File: oak_functions/loader/src/tests.rs [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+            .unwrap();
+        bencher.iter(|| wasm_handler.handle_request(&request));
+    });
+
+    // When running `cargo test` this benchmark test gets executed too, but `summary` will be `None`
+    // in that case. So, here we first check that `summary` is not empty.
+    if let Some(summary) = summary {
+        // We expect the `mean` time for loading the test Wasm module and running its main function
+        // to be less than 50ms.
+        assert!(summary.mean < 50_000.0);
```
Comment (tiziano88) [inline oak_functions/loader/src/tests.rs:84]: Nice! I think this is probably enough for what we need, and it is fine if we need to increase this number at some point, at least it will be obvious that something is taking longer. (BTW it is a bit odd that `summary.mean` returns an `f64` but it's unclear what unit that is, I guess ns, but it's not written explicitly anywhere obvious when I looked around)

**Review by tiziano88 — APPROVED** [review]

File: oak_functions/loader/src/main.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (rbehjati) [reply oak_functions/loader/src/main.rs:54]: I see. Makes sense.

File: oak_functions/loader/src/tests.rs (hunk 14 above)
Comment (rbehjati) [reply oak_functions/loader/src/tests.rs]: I agree, `example.com` is much better. Fixed. Also added a comment.

File: oak_functions/loader/src/tests.rs (hunk 15 above)
Comment (rbehjati) [reply oak_functions/loader/src/tests.rs:88]: Yes. I had to experiment with it a bit to figure out the unit. Added a comment.

**Comment by tiziano88** [comment]
Is this PR blocked by the CI failures?

**Comment by rbehjati** [comment]
> Is this PR blocked by the CI failures?

Yes. It keeps failing with `No space left on device`. Not sure how to do some diagnosis. There are no logs from the `runner` when this happens.
